### ISSUE TITLE
Using a batchsize greater than 1 when using C++ API

### ISSUE THREAD

**Issue by ppries** [tensorflow/tensorflow#3362]
I have a Tensorflow model trained in Python and frozen with the freeze_graph script. I've succesfully loaded the model in C++ and made inference for a single image. However, whenever I attempt to feed the model a batch of images in C++ I get the error:

`tensorflow/core/framework/tensor.cc:433] Check failed: 1 == NumElements() (1 vs. 16)Must have a one element tensor`

If I print the frozen graph def, the input placeholder seems to have the correct shape for batch inference:

    node {
      name: "inputs"
      op: "Placeholder"
      attr {
        key: "dtype"
        value {
          type: DT_FLOAT
        }
      }
      attr {
        key: "shape"
        value {
          shape {
            dim {
              size: 16
            }
            dim {
              size: 50
            }
            dim {
              size: 50
            }
            dim {
              size: 3
            }
          }
        }
      }
    }

I'm using the `LoadGraph` function from the Tensorflow `label_image` C++ example, and I can't see anything in that function that should limit me to one image at a time.

Is this a limitation of the C++ API or am I doing something wrong?

### Environment info
Operating System: Ubuntu 14.04

Installed version of CUDA and cuDNN: 7.5, 5.0.5

(please attach the output of `ls -l /path/to/cuda/lib/libcud*`): 

    -rw-r--r-- 1 root root   322936 Aug 15  2015 /usr/local/cuda/lib64/libcudadevrt.a
    lrwxrwxrwx 1 root root       16 Aug 15  2015 /usr/local/cuda/lib64/libcudart.so -> libcudart.so.7.5
    lrwxrwxrwx 1 root root       19 Aug 15  2015 /usr/local/cuda/lib64/libcudart.so.7.5 -> libcudart.so.7.5.18
    -rwxr-xr-x 1 root root   383336 Aug 15  2015 /usr/local/cuda/lib64/libcudart.so.7.5.18
    -rw-r--r-- 1 root root   720192 Aug 15  2015 /usr/local/cuda/lib64/libcudart_static.a
    -rwxr-xr-x 1 root root 59909104 Jul 14 11:54 /usr/local/cuda/lib64/libcudnn.so
    -rwxr-xr-x 1 root root 59909104 Jul 14 11:54 /usr/local/cuda/lib64/libcudnn.so.5
    -rwxr-xr-x 1 root root 59909104 Jul 14 11:54 /usr/local/cuda/lib64/libcudnn.so.5.0.5
    -rw-r--r-- 1 root root 58775484 Jul 14 11:54 /usr/local/cuda/lib64/libcudnn_static.a

If installed from sources, provide the commit hash: 115f5185e56c3fc4c8ada87c56651434f6be585c (r0.9)

### Steps to reproduce
1. Create and train model in python.
2. Save model and checkpoint, and use `freeze_graph` to prepare for C++.
3. Load model in C++ and feed it an input tensor with a batch size larger than 1.

### What have you tried?
1. Not using a variable batch size.
2. Sanity checking the input placeholder shape of the frozen graph.

**Comment by concretevitamin** [comment]
What is the stack trace corresponding to `tensorflow/core/framework/tensor.cc:433]`?  Can you add some LOG messages to see which node is creating issues?

**Comment by ppries** [comment]
So, I'm not that proficient at C++, but I tried using `gdb` to get something that looks like a stack trace (I added `-g -rdynamic` to the compiler flags and ran it in `gdb` with `run`:

    (gdb) bt
    #0  0x00007fffddf0fc37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
    #1  0x00007fffddf13028 in __GI_abort () at abort.c:89
    #2  0x00007fffe93a5189 in tensorflow::internal::LogMessageFatal::~LogMessageFatal() () from include/tensorflow/bazel-bin/tensorflow/libtensorflow_cc.so
    #3  0x00007fffe920e6bc in tensorflow::Tensor::CheckIsAlignedAndSingleElement() const () from include/tensorflow/bazel-bin/tensorflow/libtensorflow_cc.so
    #4  0x000000000056d5aa in tensorflow::Tensor::scalar<float> (this=0x31330d0) at include/tensorflow/tensorflow/core/framework/tensor.h:524
    #5  0x000000000056bb7c in BlockFilter::getFrame (this=0x32f1dc0, t_seconds=0) at src/BlockFilter.cpp:193
    #6  0x000000000055c5fb in BlockCamera::getFrame (this=0x22137f0, t_seconds=0) at src/BlockCamera.cpp:35
    #7  0x000000000056f81e in BlockMixDev::getFrame (this=0x36bbc80, t_seconds=0) at src/BlockMixDev.cpp:16
    #8  0x000000000057dbc8 in main (argc=4, argv=0x7fffffffe3d8) at src/sportcaster.cpp:314

As to the log messages, I'm afraid you'll have to give me a hint on how to do that!

**Comment by concretevitamin** [comment]
It appears that your code `BlockFilter::getFrame` seems to create a `Tensor`, populate it with 16 elements, and then call `tensor.scalar<float>()` on it.  

I suspect this can be fixed once you change the `getFrame()` to not call `scalar<>()`.  (If you need to access the `i`-th element, call `flat<T>()(i)`.)

**Comment by ppries** [comment]
I completely forgot about that -- I was calling `.scalar<float>` on the output tensor, which obviously won't work now that the tensor has a first dimension greater than one.

Thank you so much for the help!

**Comment by ryan-cahill** [comment]
@ppries I realize that this thread isn't exactly for asking questions like this, but the TensorFlow c++ docs and examples are so sparse that I'll take help from just about anywhere at this point. I'm attempting to load a multi-layer CNN in c++ that I've successfully trained in python, but there aren't great references to see how this works. I've attempted this many different ways, and none of them seem to be going down the correct path. If you're comfortable with it, would you be able to share your python and c++ code with me for reference?

**Comment by ppries** [comment]
Apologies for the late reply -- forgot about it over the weekend!

I'm mostly just following [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/label_image/main.cc) example to load a Python trained graph in C++. Where are you having issues?

**Comment by ryan-cahill** [comment]
Thanks for the response, but I went with the easier route of using python to serve requests to my model instead of c++ and NodeJS, which I had planned on before. The all-Python approach works great so I'm shelving the c++ version of TensorFlow until I can take more time to understand it. Again, I appreciate the response and good luck with your future projects!

**Comment by ryanjay0** [comment]
My problem is simply creating a batch of images. In that example, in ReadTensorFromImageFile it reads in an image:

```
Output image_reader = DecodeJpeg(root.WithOpName("jpeg_reader"), image_reader, tensorflow::DT_FLOAT);
auto float_caster = Cast(root.WithOpName("float_caster"), image_reader, tensorflow::DT_FLOAT);

```
Then it expands the dimension with:

`auto dim_expander = ExpandDims(root, float_caster, 0);`

but instead I'd like add more float_casters into a 4D array and pass that directly to the next line ResizeBilinear instead of the padded single image batch. 

That seems trivial but I'm having trouble with the syntax. Any suggestions?

**Comment by ryanjay0** [comment]
I'm having lots of trouble with simply getting the images in the right format to pass into the model. In that example it says:

const Tensor& resized_tensor = resized_tensor[0]
Status run_status = session->Run({{input_layer, resized_tensor}}, {output_layer}, {}, &outputs);

Which takes the return from ReadTensorFromImageFile and runs it through the model. How can I get it to take multiple images in a 4D array? Or would i do that inside ReadTensorFromImage before it subtracts the mean and divides by the scale? I'm having lots of trouble with the syntax. Please help. Thanks.

**Comment by beniz** [comment]
Look at https://github.com/beniz/deepdetect/blob/tf/src/tflib.cc#L197 it may give you the solution.

**Comment by ryanjay0** [comment]
Thank you. That worked. But I'm not sure the graph from the tensorflow label_image c++ example can handle a batch. I get this error when I tried two images:

```
E tensorflow/examples/label_batch/main.cc:344] Running model failed: Invalid argument: Input to reshape is a tensor with 4096 values, but the requested shape has 2048
	 [[Node: pool_3/_reshape = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/gpu:0"](pool_3, pool_3/_reshape/shape)]]
```

Is there a way to modify the graph to take a batch? Or am I still making a mistake with the input?

**Comment by beniz** [comment]
Read the TF FAQ, you need export / freeze the graph without specifying the batch size.

**Comment by beniz** [comment]
Hi, look at the deepdetect `tf` branch, in tflib.cc file, somewhere in the `predict` function you will find the piece of code that puts images into a batch through a combat operation, and a dedicated session. I only have limited connectivity right now otherwise I'd have the exact link for you. I'll update the message when I can.

Em.

⁣

Sent from BlueMail

​

**Comment by mbrownkinatrax** [comment]
@ppries did you even get this batch size working in c++?

**Comment by mjbk** [comment]
@ppries  Were you even able to change the batch size in C++?

**Comment by ppries** [comment]
Change batch size – no. But that should not be necessary if you freeze the graph with an unspecified graph size (e.g. a placeholder with dimensions [None, H, W, C]).

**Comment by xt479910344** [comment]
@beniz,the link https://github.com/beniz/deepdetect/blob/tf/src/tflib.cc#L197 has not been existed

**Comment by xt479910344** [comment]
@ppries ,how do you change one image to a batch of image, What changes have been made in the file of label_image/main.cc

**Comment by cuevas1208** [comment]
any luck? take a look at this [Link ](https://github.com/tensorflow/tensorflow/issues/8033) 

std::vector<TF_Tensor*> input_values;
input_values.push_back(tensor);

**Comment by jimdsouza** [comment]
@xt479910344 There you go https://github.com/beniz/deepdetect/blob/master/src/tflib.cc#L235
